modules/metastorage/README.md

```diff
@@ -0,0 +1,46 @@
+# Metastorage
+
+The module for store and access to metadata. There is module is connected with this one:
+
+- metastorage-api - the module contains classes to access to metastorage service from another components.
+
+The storage is distributed to prevent losing data and based of RAFT. That distributed storage persists on top of every cluster node, however
+most of such participants are learners and only few of them are voting ones. In typical case, the group consists of only several voted nodes
+of the cluster (amount of nodes have to be odd 3 or 5). The rest of nodes listens metadata update but does not participant in voting (RAFT 
+learners).
+
+## Threading model
+
+There is only one dedicate thread to notify watches. This is a thread with prefix metastorage-watch-executor.
+The thread is created from executor in KeyValueStorage:
+
+```java
+private ExecutorService watchExecutor = Executors.newSingleThreadExecutor(new NamedThreadFactory(NamedThreadFactory
+        .threadPrefix(nodeName, "metastorage-watch-executor"),LOG));
+```
+
+Also, the storage internally contains a two threads executor to create a snapshot:
+
+```java
+private ExecutorService snapshotExecutor = Executors.newFixedThreadPool(2,new NamedThreadFactory(NamedThreadFactory
+        .threadPrefix(nodeName, "metastorage-snapshot-executor"),LOG));
+```
+
+Both of the executors have a node name in their prefixes to distinguish to which node the particular thread belongs.
+
+### Interface methods
+
+Various operations (*get()*, *getAll()*, *invoke()*) in Metastorage manager return futures. Those futures are completed when the matched
+RAFT command completes to the Metastorage group. This result appears in RAFT client executor (prefix <NODE_NAME>%Raft-Group-Client), but the
+entire replication procedure is happened tin the RAFT threads (description of that treading model available in 
+[Raft module](../raft/README.md)).
```

Comment:
Futures are returned by a number of Metastorage Manager operations, like *get()*, *getAll()*, *invoke()*, etc. Those futures are completed when the corresponding RAFT command is completed in the Metastorage group. Although the entire replication process took place in the RAFT threads, this result appears in the RAFT client executor with prefix  <NODE_NAME>%Raft-Group-Client. See RAFT module for more information about its threading model.